packages/gatsby/src/redux/nodes.ts

```diff
@@ -419,37 +419,25 @@ function addNodeToBucketWithElemMatch(
 const binarySearch = (
   values: Array<FilterValue>,
   needle: FilterValue
-): [number, number] | undefined => {
+): number => {
   let min = 0
   let max = values.length - 1
-  let pivot = Math.floor(values.length / 2)
   while (min <= max) {
+    const pivot = min + Math.floor((max - min)/ 2)
     const value = values[pivot]
     if (needle < value) {
       // Move pivot to middle of nodes left of current pivot
       // assert pivot < max
-      max = pivot
+      max = pivot - 1
     } else if (needle > value) {
       // Move pivot to middle of nodes right of current pivot
       // assert pivot > min
-      min = pivot
+      min = pivot + 1
     } else {
-      // This means needle === value
-      // TODO: except for NaN ... and potentially certain type casting cases
-      return [pivot, pivot]
+      return pivot
     }
-
-    if (max - min <= 1) {
-      // End of search. Needle not found (as expected). Use pivot as index.
-      // If the needle was not found, max-min==1 and max is returned.
-      return [min, max]
-    }
-
-    pivot = Math.floor((max - min) / 2)
   }
-
-  // Shouldn't be reachable, but just in case, fall back to Sift if so.
-  return undefined
+  return -(min + 1)
```

Comment:
This only had one caller and I kind of struggled to understand the semantics of the previous method --- sometimes it would return `[i, i+1]` when the element was present in the array and sometimes it would return `[i, i]`. I changed it to just return a value that definitively identifies the result of the search.

Reply:
So my thought here was that this binary search would be more likely to miss than to hit. In fact, it's a very outlier edge case (caused by coercion) if it would hit, so I would want to receive the two indexes between where the pivot ended.

If this, for whatever reason, just outright failed, the fallback is to go through Sift anyways. Sift is the original implementation that we're slowly replacing / making faster. The caller is checking explicitly for `undefined` and will bail entirely on the fast indexes if this search returns it.

That said, it's not pretty and I don't think it should fail so I'm not against improving that situation.

Admittedly I'm not super fond of returning a negative number in case of a miss but I'm also aware that any alternative is just slower, including returning an array like it does now. Can't win here :p